kernel/source/messaging/ipc.c

```diff
@@ -67,172 +71,414 @@ Copyright (C) 2009		John Kelley <wiidev@kelley.ca>
+	IpcCircBuf.ReadyToSendAmount--;
+	IpcCircBuf.WaitingInBufferAmount--;
+	IpcCircBuf.HadRelaunchFlag = 0;
+	mask32(HW_IPC_ARMCTRL, ~(IPC_ARM_IX1 | IPC_ARM_IX2), (IpcCircBuf.WaitingInBufferAmount == (IPC_CIRCULAR_BUFFER_SIZE - 1) ? IPC_ARM_ACK_OUT : 0) | IPC_ARM_OUTGOING);
 }
 
-void enqueue_reply(IpcMessage* req)
+static void FlushAndSendRequest(IpcRequest *request)
 {
-	if(out_cnt >= IPC_OUT_SIZE)
+	const u32 requestCommand = request->RequestCommand;
+	if (requestCommand == IOS_IOCTL)
 	{
-		gecko_printf("IPC: OUTPUT QUEUE OVERLOAD\n");
-		return;
+		DCFlushRange(request->Data.Ioctl.InputBuffer, request->Data.Ioctl.InputLength);
+		DCFlushRange(request->Data.Ioctl.IoBuffer, request->Data.Ioctl.IoLength);
 	}
-	
-	output_queue[out_cnt] = req;
-	out_cnt++;
+	else if (requestCommand == IOS_READ)
+	{
+		DCFlushRange(request->Data.Read.Data, request->Result);
+	}
+	else if (requestCommand == IOS_IOCTLV)
+	{
+		const u32 totalArgc = request->Data.Ioctlv.InputArgc + request->Data.Ioctlv.IoArgc;
+		for(u32 i = 0; i < totalArgc; ++i)
+		{
+			DCFlushRange(request->Data.Ioctlv.Data[i].Data, request->Data.Ioctlv.Data[i].Length);
+		}
+		DCFlushRange(request->Data.Ioctlv.Data, totalArgc * sizeof(IoctlvMessageData));
+	}
+
+	IpcCircBuf.BackingArray[IpcCircBuf.PrepareToSendIndex] = request;
+	IpcCircBuf.PrepareToSendIndex = (IpcCircBuf.PrepareToSendIndex + 1) % IPC_CIRCULAR_BUFFER_SIZE;
+	IpcCircBuf.ReadyToSendAmount++;
+
+	SendIpcRequest();
 }
 
-void ipc_enqueue_reuqest(IpcMessage* req)
+static int ValidateAddressInt(const u32 addr, const u32 size)
 {
-	if(in_cnt >= IPC_IN_SIZE)
-	{
-		gecko_printf("IPC: INPUT QUEUE OVERLOAD\n");
-		return;
-	}
-	
-	input_queue[in_cnt] = req;
-	in_cnt++;
+	const u32 addrEnd = addr + size;
+	const int ret = (addr < addrEnd) && ((0x10000000 <= addr && addrEnd < 0x13620000) || (addrEnd < 0x01800000));
+	if(!ret)
+		printk("IPC: failed buf check: ptr=%08x len=%d\n", addr, size);
+	return ret;
+}
+static int ValidateAddressPtr(const void* const addr, const u32 size)
+{
+	return ValidateAddressInt((u32)addr, size);
 }
 
-void ipc_process_input(void)
+void IpcHandler(void)
 {
-	if(in_cnt == 0)
+	SetThreadPriority(0, 0x40);
+	IpcHandlerRequest.Command = IOS_INTERRUPT;
+
+	const s32 messageQueue = CreateMessageQueue((void**)IpcHandlerMessageQueueData, ARRAY_LENGTH(IpcHandlerMessageQueueData));
+	if (messageQueue < 0)
 		return;
 	
-	volatile IpcMessage* req = input_queue[in_cnt-1];
-	if(req == NULL)
-		goto exit_process;
-	
-	s32 return_value = IOS_EINVAL;	
-	u8 reply = 1;
-	/*ios_module* module = NULL;
-	DCFlushRange((u32*)req, sizeof(IpcMessage));
-	DCInvalidateRange((u32*)req, sizeof(IpcMessage));
-	
-	if(req->cmd == IOS_OPEN)
+	s32 ret = RegisterEventHandler(IRQ_IPC, messageQueue, &IpcHandlerRequest);
+	if(ret != IPC_SUCCESS)
+		return;
+
+	// enable incoming & outgoing ipc messages from PPC
+	if ((read32(HW_IPC_PPCCTRL) & (IPC_PPC_IY1 | IPC_PPC_IY2)) == 0)
+		write32(HW_IPC_PPCCTRL, 0);
```

Comment:
ios has something different here. it doesn't write 0, it does and & with 0x30
```
  if ((_HW_IPC_PPCCTRL & 0x30) == 0) {
    _HW_IPC_PPCCTRL = _HW_IPC_PPCCTRL & 0x30;
  }
```

Reply:
well, this was an "optimization" because i doubted the read would change in the microseconds it takes to do the AND, but better be safe than sorry. I don't actually know how these flags work much.

Reply:
its more a case of if other bits are set. it only checks the 0x10 & 0x20 bits, but if lower bits were set they would get lost... i think. now im doubting myself because of the & read.
crap ye, write 0 has the same effect.